internal/controllers/tlsedge_controller.go

```diff
@@ -0,0 +1,374 @@
+		Recorder: r.Recorder,
+
+		kubeType: "v1alpha1.TLSEdge",
+		statusID: func(cr *ingressv1alpha1.TLSEdge) string { return cr.Status.ID },
+		create:   r.create,
+		update:   r.update,
+		delete:   r.delete,
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ingressv1alpha1.TLSEdge{}).
+		Watches(
+			&source.Kind{Type: &ingressv1alpha1.IPPolicy{}},
+			handler.EnqueueRequestsFromMapFunc(r.listTLSEdgesForIPPolicy),
+		).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=ingress.k8s.ngrok.com,resources=tlsedges,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=ingress.k8s.ngrok.com,resources=tlsedges/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ingress.k8s.ngrok.com,resources=tlsedges/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.1/pkg/reconcile
+func (r *TLSEdgeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	return r.controller.reconcile(ctx, req, new(ingressv1alpha1.TLSEdge))
+}
+
+func (r *TLSEdgeReconciler) create(ctx context.Context, edge *ingressv1alpha1.TLSEdge) error {
+	if err := r.reconcileTunnelGroupBackend(ctx, edge); err != nil {
+		return err
+	}
+
+	// Try to find the edge by the backend labels
+	resp, err := r.findEdgeByBackendLabels(ctx, edge.Spec.Backend.Labels)
+	if err != nil {
+		return err
+	}
+
+	if resp != nil {
+		return r.updateEdgeStatus(ctx, edge, resp)
+	}
+
+	// No edge has been created for this edge, create one
+	r.Log.Info("Creating new TLSEdge", "namespace", edge.Namespace, "name", edge.Name)
+	resp, err = r.NgrokClientset.TLSEdges().Create(ctx, &ngrok.TLSEdgeCreate{
+		Hostports:   edge.Spec.Hostports,
+		Description: edge.Spec.Description,
+		Metadata:    edge.Spec.Metadata,
+		Backend: &ngrok.EndpointBackendMutate{
+			BackendID: edge.Status.Backend.ID,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	r.Log.Info("Created new TLSEdge", "edge.ID", resp.ID, "name", edge.Name, "namespace", edge.Namespace)
+
+	if err := r.updateEdgeStatus(ctx, edge, resp); err != nil {
+		return err
+	}
+
+	return r.updateIPRestrictionRouteModule(ctx, edge, resp)
+}
+
+func (r *TLSEdgeReconciler) update(ctx context.Context, edge *ingressv1alpha1.TLSEdge) error {
+	if err := r.reconcileTunnelGroupBackend(ctx, edge); err != nil {
+		return err
+	}
+
+	resp, err := r.NgrokClientset.TLSEdges().Get(ctx, edge.Status.ID)
+	if err != nil {
+		// If we can't find the edge in the ngrok API, it's been deleted, so clear the ID
+		// and requeue the edge. When it gets reconciled again, it will be recreated.
+		if ngrok.IsNotFound(err) {
+			r.Log.Info("TLSEdge not found, clearing ID and requeuing", "edge.ID", edge.Status.ID)
+			edge.Status.ID = ""
+			//nolint:errcheck
+			r.Status().Update(ctx, edge)
```

Comment:
check the error here, instead of masking it?

Reply:
:+1: Will fix in TCP as well.